mailchimp_auth/mailchimp.py

```diff
@@ -0,0 +1,185 @@
+from django.conf import settings
+import email_normalize
+
+import mailchimp_marketing as MailchimpMarketing
+from mailchimp_marketing.api_client import ApiClientError
+
+
+class MailchimpAPI(object):
+    '''
+    Wrapper for supporter methods:
+    https://mailchimp.com/developer/marketing/api/list-members/
+    '''
+    LIST_ID = settings.MAILCHIMP_LIST_ID
+    API_KEY = settings.MAILCHIMP_API_KEY
+    SERVER = settings.MAILCHIMP_SERVER
+
+    def _has_valid_email(self, supporter, email_address):
+        '''
+        Determine whether a supporter has a valid contact matching the given
+        email address.
+        '''
+        email_valid = (email_normalize.normalize(supporter["email_address"]) == email_normalize.normalize(email_address) and
+                        supporter['status'] != 'unsubscribed')
+
+        if email_valid:
+            return True
+
+        return False
+
+    def put_supporter(self, user):
+        '''
+        Add or update supporter.
+        '''
+
+        payload = {
+            "email_address": user.email,
+            "status": "subscribed",
+            "merge_fields": {
+                "FNAME": user.first_name,
+                "LNAME": user.last_name,
+            }
+        }
+
+        subscriber = payload["email_address"]
+
+        try:
+            client = MailchimpMarketing.Client()
+            client.set_config({
+                "api_key": self.API_KEY,
+                "server": self.SERVER
+            })
+
+            response = client.lists.set_list_member(self.LIST_ID, subscriber, payload)
+
+        except ApiClientError as error:
+            print("Error: {}".format(error.text))
+
+        return response  # A dict representing the member that has been added
+
+    def get_supporter(self, email_address, allow_invalid=False):
+        '''
+        Return the supporter with an exactly matching email address that is valid.
+        '''
+
+        try:
+            client = MailchimpMarketing.Client()
+            client.set_config({
+                "api_key": self.API_KEY,
+                "server": self.SERVER
+            })
+
+            response = client.searchMembers.search(query=email_address, fields=["exact_matches"])
+
+        except ApiClientError as error:
+            print("Error: {}".format(error.text))
+
+        if response['exact_matches']['total_items'] == 1:
+            supporter = response['exact_matches']['members'][0]
+
+            if allow_invalid:
+                return supporter
+
+            elif self._has_valid_email(supporter, email_address):
```

Comment:
let's have this check before we make the api call?

Reply:
How so? I think this might be dependent on making the api call, since `supporter` comes from the response and `has_valid_email` checks to see if the status of that supporter in mailchimp is anything other than `unsubscribed`.

Sidenote: there's another check that's in `has_valid_email` that feels like it could be redundant so I could get rid of that method for simplicity's sake and just change these lines to something like
```
if allow_invalid or supporter['status'] != 'unsubscribed':
    return supporter
```